url46a88cc932_toby1984_tinyc_tgz-pJ8-de_codesourcery_dcpu16_compiler_CompilerJ8/src/de/codesourcery/dcpu16/codegeneration/CPUModel.java

```diff
@@ -0,0 +1,449 @@
+        {
+        	if ( isRegisterIndirection( currentCmd ) && isLoadAddress( queue ) )
+        	{
+        		final MemoryTarget oldTarget = (MemoryTarget) queue.operand2Target;
+        		final MemoryTarget newTarget = new MemoryTarget( oldTarget.label , true );
+                queue = new AssemblyCommand( OpCode.SET , currentCmd.operand1Target , newTarget ).withComment( currentCmd.comment );
+                return;
+        	}
+        }
+        
+        /*
+         * Combine:
+         * 
+         * SET $2 , $1
+         * SET $3 , $2
+         * 
+         * into
+         * 
+         * SET $3 , $1
+         * 
+         * TODO: This will break stuff if something relies on the values availability in A 
+         */
+        if ( currentCmd.isCommand( OpCode.SET ) && queue.isCommand( OpCode.SET ) ) 
+        {
+            final String loadSource = queue.operand2;
+            final String loadDestination = queue.operand1;
+            final String storeSource = currentCmd.operand2;
+            if ( equals( storeSource, loadDestination ) ) 
+            {
+                final String storeDestination = currentCmd.operand1;                
+                AssemblyCommand newCmd = new AssemblyCommand( OpCode.SET , storeDestination , loadSource ).withComment( currentCmd.comment );
+                if ( DEBUG_PEEPHOLE ) {
+                    System.err.println("; PEEPHOLE: "+queue+" + "+currentCmd+" => "+newCmd);
+                }
+                queue = newCmd;
+                return;
+            }
+        } 
+        
+        internalOutput( queue );
+        queue = currentCmd;
+    }
+    
+    private static boolean isRegisterIndirection(AssemblyCommand currentCmd) 
+    {
+		if ( currentCmd.isCommand( OpCode.SET) &&
+			 currentCmd.operand1Target instanceof RegisterTarget &&
+			 currentCmd.operand2Target instanceof RegisterTarget ) 
+		{
+			RegisterTarget dest = (RegisterTarget) currentCmd.operand1Target;
+			RegisterTarget source = (RegisterTarget) currentCmd.operand2Target;
+			return source.getRegister() == dest.getRegister() &&
+				   source.isIndirect() && ! dest.isIndirect();
+		}
+		return false;
+    }
+    
+    private static boolean isLoadAddress(AssemblyCommand currentCmd) 
+    {
+		if ( currentCmd.isCommand( OpCode.SET) &&
+			 currentCmd.operand1Target instanceof RegisterTarget &&
+			 currentCmd.operand2Target instanceof MemoryTarget ) 
+		{
+			RegisterTarget dest = (RegisterTarget) currentCmd.operand1Target;
+			MemoryTarget source = (MemoryTarget) currentCmd.operand2Target;
+			return ! dest.isIndirect() && ! source.isIndirect();
+		}
+		return false;
+    }    
+    
+    private final void internalOutput(AssemblyCommand cmd) 
+    {
+        writer.codeSegment();
+        writer.write( cmd.toString()+"\n" );
+    }
+
+    private static boolean equals(String s1,String s2) 
+    {
+        if ( s1 == null || s2 == null ) {
+            return s1 == s2;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.